webui/src/lib/components/repository/repoOnboardingChecklistSlider.tsx

```diff
@@ -0,0 +1,200 @@
+  const [hasPermissions, setHasPermissions] = useState(false);
+  const [steps, setSteps] = useState<Array<StepsWithStatus>>([]);
+  const [checklistTop, setChecklistTop] = useState(0);
+  const checklistBodyRef = useRef<HTMLDivElement>(null);
+
+  const toggleShowChecklistHandler = useCallback(() => {
+    showChecklist(!show);
+  }, [showChecklist, show]);
+
+  const handleOffcanvasEnter = useCallback(() => {
+    // make sure the checklist is centered one it's mounted to the DOM
+    // (this is needed because the checklist is rendered inside an offcanvas)
+    // the formula is: topOfChecklist = (viewportHeight - checklistHeight) / 2
+    if (checklistBodyRef.current) {
+      const viewportHeight = window.innerHeight;
+      const newChecklistTop =
+        (viewportHeight -
+          checklistBodyRef.current.getBoundingClientRect().height) /
+        2;
+      setChecklistTop(newChecklistTop);
+    }
+  }, [checklistBodyRef.current, setChecklistTop]);
+
+  useEffect(() => {
+    const resolveIsSampleRepo = async () => {
+      const isDemo = await isSampleRepo(repo.id);
+      setIsSampleRepoState(isDemo);
+    };
+    if (repo) {
+      resolveIsSampleRepo();
+    }
+  }, [repo]);
+
+  useEffect(() => {
+    const resolveHasPermissions = async () => {
+      const hasPerms = await canUseRepoOnboarding(repo.id);
+      setHasPermissions(hasPerms);
+    };
+    if (repo) {
+      resolveHasPermissions();
+    }
+  }, [repo]);
+
+  useEffect(() => {
+    const resolveSteps = async () => {
+      const steps = getRepoOnboardingSteps(
+        repo.id,
+        repo.default_branch,
+        blockstoreType,
+        navigate
+      );
+      const stepsWithStatus = steps.map((step) => ({
+        ...step,
+        isCompletedValue: false,
+        isLoading: true,
+      }));
+      setSteps(stepsWithStatus);
+    };
+    if (repo && hasPermissions && !isSampleRepoState) {
+      resolveSteps();
+    }
+  }, [repo, hasPermissions, isSampleRepoState, blockstoreType, navigate]);
+
+  useEffect(() => {
+    const stepStatusResolver = async (
+      stepId: string,
+      stepCompleteFn: (repoId: string) => Promise<boolean>
+    ) => {
+      const status = await stepCompleteFn(repo.id);
+      setSteps((steps) => {
+        const stepIndex = steps.findIndex((step) => step.id === stepId);
+        const newSteps = [...steps];
+        newSteps[stepIndex] = {
+          ...newSteps[stepIndex],
+          isCompletedValue: status,
+          isLoading: false,
+        };
+        return newSteps;
+      });
+    };
+
+    if (steps.length > 0) {
+      steps
+        .filter((step) => step.isLoading)
+        .forEach((step) => {
+          if (step.showStep()) {
+            stepStatusResolver(step.id, step.isCompleted);
+          }
+        });
+    }
+  });
+
+  const [finishedSteps, totalSteps] = useMemo(() => {
+    const totalSteps = steps.reduce(
+      (n, current) => n + (current.showStep() ? 1 : 0),
+      0
+    );
```

Comment:
This is really hard to read.  Consider using any less-general operator than reduce, even if it potentially generates some garbage.  E.g.
```suggestion
    const totalSteps = steps.filter(current => current.showStep()).length;
```

Reply:
Good call! Done.